Update entry.cpp and multilineentry.cpp on Windows

norepimorphism opened the pull request:
I'm not sure if it is better etiquette to make individual PRs for each of my suggested changes, but there are multiple small changes so I decided to group them into a single PR.

https://github.com/norepimorphism/libui-ng/blob/7b9380b17692807bb5a4164213d01018ab2f7c6c/windows/entry.cpp#L16-L23

Here, I made a small refactor to make the branches taken a little more clear. I made the same change in `multilineentry.cpp`.

https://github.com/norepimorphism/libui-ng/blob/7b9380b17692807bb5a4164213d01018ab2f7c6c/windows/entry.cpp#L71

Here, I replaced a `PostMessageW` call with the corresponding macro: in this case, `Edit_SetSel`. I believe switching from `PostMessageW` to macros where possible was mentioned previously as an issue.

https://github.com/norepimorphism/libui-ng/blob/7b9380b17692807bb5a4164213d01018ab2f7c6c/windows/entry.cpp#L89-L90

This should be equivalent to the original code but is a little less verbose.

https://github.com/norepimorphism/libui-ng/blob/7b9380b17692807bb5a4164213d01018ab2f7c6c/windows/multilineentry.cpp#L84-L107

I made a few changes here. Mainly, I addressed the comments left previously by preserving the current text selection after appending text. I also made a few macro replacements where possible, and I introduced a new function `windowTextLen` in `text.cpp` that returns the text length of an edit control. I noticed we were querying text length manually in some places, so I decided to make a dedicated function.

szanni reviewed (commented):
Thanks for taking a jab at fixing this.

In general it would be helpful to have the description of what the PR does at the top of the comment instead of at the bottom. I had difficulty figuring out what the PR is trying to solve. And maybe adding a comment on how to test (tester > page12).

Apart from that the general programming style is early return, the `onWM_COMMAND` functions are better the way they are currenty (less indentation, easier flow).

And I am unsure if the wrapping `windowTextLen` makes the code better of worse. It is a one liner and I would like to avoid starting to wrap all win32 API calls.

Otherwise good job, it seems to work from my limited testing.

norepimorphism commented on windows/entry.cpp:
```diff
@@ -13,13 +13,13 @@ static BOOL onWM_COMMAND(uiControl *c, HWND hwnd, WORD code, LRESULT *lResult)
-	(*(e->onChanged))(e, e->onChangedData);
-	*lResult = 0;
-	return TRUE;
+	} else {
```
Yes, that should probably be added to the style guide.

norepimorphism commented on windows/entry.cpp:
```diff
@@ -86,13 +86,8 @@ int uiEntryReadOnly(uiEntry *e)
-		ro = (WPARAM) TRUE;
-	if (SendMessage(e->hwnd, EM_SETREADONLY, ro, 0) == 0)
-		logLastError(L"error making uiEntry read-only");
+	if (Edit_SetReadOnly(e->hwnd, readonly != 0) == 0)
```
The original code *is* quite strange. My code here should be semantically equivalent to the original, but your change is fine also.

norepimorphism commented:
- I removed `windowTextLen`, replacing such calls with `SendMessageW(hwnd, WM_GETTEXTLENGTH, 0, 0)`;
- I reverted the change regarding early exits;
- I replaced `LRESULT n` with `LRESULT l`;
- I replaced `readonly != 0` with `readonly`; and
- I replaced snake case identifiers with camel case ones

norepimorphism commented:
In a minute I will push the comment changes you mentioned.

norepimorphism commented:
Also, correct me if I'm wrong, but I believe toggling `inhibitChanged` is unnecessary here:
https://github.com/libui-ng/libui-ng/blob/8037628a7c3b7e798d9f648955efdc2e2554664c/windows/multilineentry.cpp#L71-L82

`uiWindowsSetWindowText` invokes the Windows function `SetWindowTextW`, and [the Microsoft `EN_CHANGE` documentation](https://learn.microsoft.com/en-us/windows/win32/controls/en-change#remarks) states that an `EN_CHANGE` notification is not emitted when `WM_SETTEXT` is sent to a multiline edit control (which is what `SetWindowTextW` does).

That isn't to say we need to remove the `inhibitChanged` code; maybe it's better to be safe than sorry. I just want to point out that I think it's unnecessary, and the comment `// doing this raises an EN_CHANGED` (sic) may be misleading.

cody271 reviewed (commented):
> I'm not sure if it is better etiquette to make individual PRs for each of my suggested changes, but there are multiple small changes so I decided to group them into a single PR.

A single PR is good. As long as the patch is a logical grouping of changes like you have here in this PR. One minor issue I do see here:

```
* Update entry.c and multilineentry.c on Windows
* Address comments to PR libui-ng#146
* Push comment changes
```
^ The git commit log doesn't really stand on its own. Without going through the GitHub discussion or re-reviewing the code, I'd have no idea what these changes actually do.

For example the git commit message could be `Refactor entry.c and multilineentry.c on Windows` instead of `Update ...`, to indicate that there is no change in functionality intended. Only code quality improvements.

norepimorphism commented:
I was planning to squash the commits but was going to wait first and see if there were any further comments. If not, I will go ahead and squash them with something ala the commit message you have written.

To be pedantic, this slightly alters the functionality of multiline text appending slightly&mdash;it preserves the user's selection, which was previously overwritten&mdash;hence the wording is vague. But 'refactor' is good enough.

szanni commented:
I would definitely appreciate clearer commit messages/logs in general and enforcing such as a project policy.
I've run into multiple code sections in the past where I could not figure out things. Sadly git blame would only give me `More stuff fixed`, or even horrors like `Bleh`.

> That isn't to say we need to remove the `inhibitChanged` code; maybe it's better to be safe than sorry. I just want to point out that I think it's unnecessary, and the comment `// doing this raises an EN_CHANGED` (sic) may be misleading.

This is a perfect test case for the unit tests. I can see if I find some time the coming days. The windows docs can sadly not be trusted, at all! This is one of the things I learned while fixing some things in wine.

cody271 commented:
> this slightly alters the functionality of multiline text appending slightly—it preserves the user's selection, which was previously overwritten

Then ideally this part of the PR would be in a separate commit. Keep the refactor commit just about refactoring if possible.

> I would definitely appreciate clearer commit messages/logs in general and enforcing such as a project policy.
> I've run into multiple code sections in the past where I could not figure out things.
>
`Did some stuff` is never an acceptable commit message! 😛

szanni reviewed (approved):
Code looks good to me.
Tested, behavior now matches unix and darwin. I did not find any regressions.

If you could squash/separate the refactoring and selection fix this would be perfect. Good work

szanni reviewed (approved):
Signed off
